engine/schema/src/com/cloud/host/dao/HostDaoImpl.java

```diff
@@ -1194,4 +1178,26 @@ public HostVO findByIp(final String ipAddress) {
         sc.setParameters("type", type);
         return listBy(sc);
     }
-}
+
+    String sqlFindHostConnectedToSnapshotStoragePoolToExecuteCommand = "select h.id from snapshots s join volumes v on v.id = s.volume_id "
+            + "join storage_pool pool on pool.id = v.pool_id join cluster c on pool.cluster_id = c.id join host h on h.cluster_id = c.id "
```

Comment:
This method will not work in scenarios when volume or storage pool associated to snapshot is deleted. Also there is no need to consider storage pool as only secondary storage is involved. Any random hypervisor host in zone will do. Scenarios where storage pool is involved are already handled in core code and don't go via command delegation route. Also don't use sql queries instead write queries using Daos.

Reply:
I am using a DAO (Data access object), did you mean using a JPA queries (with ACS' home made JPA implementation)?

But if volume is deleted, how can we find its snapshots? I mean, aren't the snapshots of a volume deleted as well?

Are you sure that we use only the files in the secondary storage to build the template from a snapshot? I was under the impression that it was being used the primary storage as well.

Reply:
Yes, we only use secondary storage to build the template from snapshot. This might not be applicable to managed store. But that's also available to all hosts in zone so any random host in a zone will do.

Also we can delete volume without deleting its snapshot (backup). We only delete VM snapshots when we delete VM.

Reply:
@anshul1886  I tested this solution here to recover a volume from a snapshot of a volume that has been deleted, and it works. As you can see I am not filtering the volumes that have been deleted. 

I also tested what you said that we only use the secondary storage to build the volume from snapshots, and it is right. This means, any XenServer of the Zone can execute this command then. I will update my select to reflect that.

Regarding JPA, I have mixed feeling about the JPA implementation ACS uses. It is not a standard one and it requires too much code to achieve the same thing we do with SQL. Also, It does not seem to be a standard in ACS to do everything with JPA. I would rather use Spring-data with a commercial JPA implementation (EclipseLink, Hibernate, OpenJPA), but this requires some structural changes.

Reply:
@rafaelweingartner SQL queries are hard to follow and is not recommended way. See https://github.com/apache/cloudstack/pull/2317 for example.